app/controllers/application_controller.rb

```diff
@@ -59,30 +61,26 @@ def valid_help_actions
     @valid_help_actions ||= %w[signed_in create].freeze
   end
 
-  def confirm_two_factor_setup
-    user = request.env["warden"].user
-
-    if Rails.configuration.enable_enhanced_2fa_experience
-      return if user.nil?
-
-      if user.second_factor_method.present?
-        # Handled by Devise's controller.
-        return if user.second_factor_method == "app"
+  def tfa_feature_flag?
+    Rails.configuration.enable_enhanced_2fa_experience
+  end
 
-        # Handled by us.
-        redirect_to users_two_factor_authentication_email_path
-        return
-      end
+  def choose_two_factor_method
+    return if user.nil? || !user.second_factor_method.nil?
```

Comment:
for all these `check X first then check X.Y` scenarios I think it might be easier to use the safe operator (`user&.second_factor_method`) or any of the [techniques described here](http://mitrev.net/ruby/2015/11/13/the-operator-in-ruby/). I quite like `.try` as well which is ActiveRecord specific but is nice and clear.

Reply:
Rubocop might concur with `return unless user.try(:second_factor_method).present?` (the `.present` isn't necessary but it makes it easier to read)

Reply:
This one is not quite the same. ```User.nil?``` signifies the case that the user hasn't signed in - the second bit of the expression is a logged in user not having a ```second_factor_method``` set. Say the user hasn't logged in then the second bit is making this expression always true if I use the safe operator. I think the ```unless``` case in your second comment is harder to read

Reply:
> Say the user hasn't logged in then the second bit is making this expression always true if I use the safe operator. 

Is always false. `nil.try(:anything)` -> `nil`.

Happy to stick to your `if` version.